### ISSUE TITLE
Feature Request: Retain word metadata or output in original sort order

### ISSUE THREAD

**Issue by bwbug** [sts10/tidy#8]
First of all, great work!  I am just getting started with passphrases and word lists, and expect this tool to be a tremendously useful resource.

Let me try to explain my request.  Word list creation may include additional metadata about each word.  For example, EFF considered both ["prevalence"](http://crr.ugent.be/archives/2045) and ["concreteness"](http://crr.ugent.be/archives/1330) when designing their Long Wordlist.  In general, one might want to create a word list that contains the first _N_ words sorted according to some metadata quantity (e.g., a weighted score of "prevalence" and "concreteness").  However, the default behavior of returning the new word list sorted alphabetically makes this difficult.

Thus, a simple enhancement would be to include an option to return the new word list in its original sort order, instead of being sorted alphabetically.  A more involved functionality enhancement would be to accept input word lists that contain one or more comma-separated values following each word, and preserve these values unchanged in the final output (this would allow the user to manually sort or filter the resulting list after it has been processed by Tidy).

**Comment by sts10** [comment]
Thanks for the compliment! Hope you find the tool useful. 

> ... a simple enhancement would be to include an option to return the new word list in its original sort order, instead of being sorted alphabetically.

Love this idea. I decided to keep the alphabetical sort by default, but allow users to override it with a new option: 

```
    -O, --no-alpha
            Do NOT sort outputted list alphabetically. Preserves original list order. Note that
            duplicates lines and blank lines will still be removed
```

> A more involved functionality enhancement would be to accept input word lists that contain one or more comma-separated values following each word, and preserve these values unchanged in the final output

I'm a bit confused by this. Currently, a list like: 
```
and,1000
toucan,4
needless,50
```
would keep the commas and integers if you ran it through Tidy without any options. 

I'd say that being able to sort words like the above by the after-comma integer is a bit outside of the scope of Tidy, but I maybe could be convinced...

**Comment by sts10** [comment]
Totally understand the hesitancy to install Rust just to use Tidy. 

> any chance of you distributing an executable for Windows?

Unfortunately I don't have much experience with Rust binaries, but as an experiment I created a release with a binary here: https://github.com/sts10/tidy/releases/tag/v0.2.30

I give it about a 25% chance ti works for Windows/you, but figured it was worth a shot. I'll keep poking around for a "How to make a Rust binary for Windows" blog post...

**Comment by sts10** [comment]
Ah, sorry. 

Let me know if you get a binary working and I'll gladly host it on this repo!

**Comment by sts10** [comment]
Ah, yes, now I understand, thanks to the `tidy -Om 3 words.txt` example. It'd be like an --ignore-but-preserve-after-first comma option, haha.

Given the way I laid out the code, this would be difficult to implement without making a big mess in the code. But now that I understand the request i can mull it over a bit. Maybe there is a creative solution.

**Comment by sts10** [comment]
Pushed a basic -K keep-metadata option. 

1.) I found that it would be tricky to have it work with otpions that **modify** words, like "to lowercase". This is because of the duplication problem. 

```
The,976213
the,230947
```
It's unclear which metadata to keep if user chooses to lowercase all words.

2.) I didn't figure out how to make keep-metadata work with option like remove prefix words and remove suffix words. Will have to think about it some more. 

3.) Likewise with the list attributes. Maybe this will be fixable.

**Comment by sts10** [comment]
> Will you be using my idea of letting the -T, -S and -C flags do double duty to specify which delimiter to use for the --ignore-metadata option?

Ha, I thought of this! `-g` actually accepts a string, allowing the user to define any delimiter they want. So for example, for the Wiki list you'd run `tidy -O -m 3 -g ' ' enwiki-20210820-words-frequency.txt` 

I guess I should overhaul the -t -s -c flags to just be one that takes a delimiting string like -g does.

**Comment by sts10** [comment]
Yep, I was thinking the same this today. 

> accept a user-defined string (like you currently do with -g)

Yep, I just consolidated all of those delete option to just two: 

```
    -d, --delete-after <DELETE_AFTER_DELIMITER>
            Delete all characters after given delimiter (including the delimiter). Maximum of one
            character. Use 't' for tab and 's' for space

    -D, --delete-through <DELETE_THROUGH_DELIMITER>
            Delete all characters up to and including given delimiter. Maximum of one character. Use
            't' for tab and 's' for space
```

I like lowercase for forward, uppercase for backwards, as that's how some commands in Vim work.

Next, I'll work on a `-K` flag that ignores metadata _before_ the word. Sound good?

> So, where the help says -g, --ignore-metadata <IGNORE_METADATA>, should <IGNORE_METADATA> actually be <DELIMITER>?

Whoops -- fixed.

**Comment by sts10** [comment]
> So will the two versions of ignore be -k and -K (or -g and -G)?

Yes. And I'm switching to =g and -G... that -K reference above was a mistake on my part.

> However, what happens currently if there are repeats of the single-character delimiter?

A bit by chance, but Tidy splits on the first space, then trims any leading whitespace on the word, which eliminates the second space in this case. However we wouldn't be so lucky if the delimiter with two commas or two of some other non-whitespace character, given the current set up.

>  If you don't treat repeated delimiters as if they were a single delimiter, then this would be even more reason to re-define the backwards actions (-D and -K) to delete/ignore up to and including the last appearance of the delimiter (a notion I'm partial to anyway, to create symmetry with the forward actions -d and -k).

Oof, I'll have to think more on this some other time.

**Comment by sts10** [comment]
Phew, think I got `-G` working! Need to do some more testing, maybe tomorrow.

**Comment by sts10** [comment]
> So, currently, repeated whitespace delimiters are in effect treated as a single delimiter, but repeated non-whitespace delimiters are not, correct? In that case, I think it would be OK just to stick with your current paradigm of splitting on the first delimiter (whether working forwards or backwards). Making a symmetric implementation (delete ignore before first vs. after last) is less essential if repeated whitespace is automatically trimmed; the multiple-commas scenario is an edge case.

Agreed. Keeping it where it is for now. 

> My suggestion for the option names and help text would be as follows:...

This looks good. I've made these changes with only a minor tweak. I don't _love_ the from/through language, but I don't have an alternative that I like better at the moment. 

> What happens when these options are used together with word modification options? Are those incompatible options just ignored, or does it produce incorrect results (or worse, crash)?

Technically it crashes? It's a `panic!` call, so Rust ensures that there's no real computer issues. Plus, Rust's `panic!` macro allows you to write a message, which I did. Thus, here's what happens if you try to use `-G` and `-g`:

```
❯ tidy -G , -g s --samples l/delimit_test.txt 
thread 'main' panicked at 'Can't ignore metadata on both sides currently', src/lib.rs:169:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
A similar message is displayed if the user tries to use on the g commands with an incompatible modification option:
```
❯ tidy -G , -l l/delimit_test.txt 
thread 'main' panicked at '--ignore-starting-metadata option does not work with one of the other options you selected. Please reconsider. Exiting', src/main.rs:281:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I can do some work to make the message display a bit cleaner/prettier and/or re-write it, but I think it's fine for now. Nothing catastrophic happens.

I will note that users, thankfully, _can_ use `-D` and `-d` together. For example:
```
mA1,word1 mB1
mA2,word2 mB2
mA3,word3 mB3
```
We can run `tidy -D , -d s l/delimit_test.txt` and we get:
```
word1
word2
word3
```
I think this is a feasible use-case, so glad it works as expected.

> Let me know when you think the latest version is sufficiently stable that I should update the Windows binary.

I'm not sure! I think we've got the `-G` and `-g` options in a good place, but this was a lot of changes in a short period of time, so we'll likely catch things in the next few days. Maybe hold for a bit?

**Comment by sts10** [comment]
> "before/after"? 

Yeah, I think that makes more sense to me. I've made the (tedious) change. 

> Interesting! What would happen for data of the form `A,B` or `A,B,C` if you try `tidy -D , -d ,`?

Huh, this is tricky. With that command, `A,B,C` becomes `B` (I fixed the order of delete calls to ensure this logical outcome). `A,B` becomes `B` -- not sure if that's good or bad!

**Comment by sts10** [comment]
FYI I fixed two bugs this morning. The important one is that we have to do word modifications **BEFORE** we do word removals. This is because it's possible for a word to modified such that, for example, it dips below a user-requested word-length minimum. If we did removals and then modifications, as I had in the last few versions, a word below the requested word-length minimum could make it into the final list. 

The second bug was a dumb coding mistake with one of the -g flags. 

Both of these bugs are (hopefully) fixed in version 0.2.51 (you can always check the version number in line 3 of the `Cargo.toml` file). Would obviously recommend you not create a Windows binary of versions earlier than 0.2.51.

**Comment by sts10** [comment]
I think we can close this one? Great work!
